File: puzzle.go
// puzzle.go
package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"math/rand"
	"os"
	"os/exec"
	"runtime"
	"strconv"
	"strings"
	"time"
)

const (
	reset  = "\033[0m"
	red    = "\033[91m"
	green  = "\033[92m"
	yellow = "\033[93m"
	blue   = "\033[94m"
	magenta= "\033[95m"
	cyan   = "\033[96m"
	white  = "\033[97m"
	gray   = "\033[90m"
	bold   = "\033[1m"
)

func colorize(text, color string) string {
	return color + text + reset
}

type Puzzle struct {
	size      int
	board     [][]int
	emptyR, emptyC int
	moves     int
	startTime time.Time
}

func NewPuzzle(size int, state [][]int) *Puzzle {
	p := &Puzzle{size: size, moves: 0}
	if state != nil {
		p.board = make([][]int, size)
		for i := 0; i < size; i++ {
			p.board[i] = make([]int, size)
			copy(p.board[i], state[i])
			for j := 0; j < size; j++ {
				if p.board[i][j] == 0 {
					p.emptyR, p.emptyC = i, j
				}
			}
		}
	} else {
		p.board = make([][]int, size)
		for i := 0; i < size; i++ {
			p.board[i] = make([]int, size)
			for j := 0; j < size; j++ {
				p.board[i][j] = i*size + j + 1
			}
		}
		p.board[size-1][size-1] = 0
		p.emptyR, p.emptyC = size-1, size-1
	}
	return p
}

func (p *Puzzle) render() string {
	var lines []string
	lines = append(lines, colorize("┌"+strings.Repeat("─", p.size*4-1)+"┐", gray))
	for i := 0; i < p.size; i++ {
		line := colorize("│", gray)
		for j := 0; j < p.size; j++ {
			val := p.board[i][j]
			if val == 0 {
				line += "   "
			} else {
				colors := []string{red, green, yellow, blue, magenta, cyan, white, gray}
				col := colors[val%7]
				line += colorize(fmt.Sprintf("%2d ", val), col)
			}
		}
		line += colorize("│", gray)
		lines = append(lines, line)
	}
	lines = append(lines, colorize("└"+strings.Repeat("─", p.size*4-1)+"┘", gray))
	return strings.Join(lines, "\n")
}

func (p *Puzzle) isSolved() bool {
	target := 1
	for i := 0; i < p.size; i++ {
		for j := 0; j < p.size; j++ {
			if i == p.size-1 && j == p.size-1 {
				if p.board[i][j] != 0 {
					return false
				}
			} else {
				if p.board[i][j] != target {
					return false
				}
				target++
			}
		}
	}
	return true
}

func (p *Puzzle) isSolvable() bool {
	flat := []int{}
	for i := 0; i < p.size; i++ {
		for j := 0; j < p.size; j++ {
			if p.board[i][j] != 0 {
				flat = append(flat, p.board[i][j])
			}
		}
	}
	inversions := 0
	for i := 0; i < len(flat); i++ {
		for j := i + 1; j < len(flat); j++ {
			if flat[i] > flat[j] {
				inversions++
			}
		}
	}
	if p.size%2 == 1 {
		return inversions%2 == 0
	} else {
		if p.emptyR%2 == 0 {
			return inversions%2 == 1
		} else {
			return inversions%2 == 0
		}
	}
}

func (p *Puzzle) shuffle(moves int) {
	dirs := [][2]int{{0, 1}, {0, -1}, {1, 0}, {-1, 0}}
	for i := 0; i < moves; i++ {
		r, c := p.emptyR, p.emptyC
		rand.Shuffle(len(dirs), func(i, j int) { dirs[i], dirs[j] = dirs[j], dirs[i] })
		for _, d := range dirs {
			nr, nc := r+d[0], c+d[1]
			if nr >= 0 && nr < p.size && nc >= 0 && nc < p.size {
				p.board[r][c], p.board[nr][nc] = p.board[nr][nc], p.board[r][c]
				p.emptyR, p.emptyC = nr, nc
				break
			}
		}
	}
	p.moves = 0
	p.startTime = time.Time{}
}

func (p *Puzzle) move(dr, dc int) bool {
	r, c := p.emptyR, p.emptyC
	nr, nc := r+dr, c+dc
	if nr >= 0 && nr < p.size && nc >= 0 && nc < p.size {
		p.board[r][c], p.board[nr][nc] = p.board[nr][nc], p.board[r][c]
		p.emptyR, p.emptyC = nr, nc
		p.moves++
		if p.startTime.IsZero() {
			p.startTime = time.Now()
		}
		return true
	}
	return false
}

func clearScreen() {
	cmd := exec.Command("clear")
	if runtime.GOOS == "windows" {
		cmd = exec.Command("cmd", "/c", "cls")
	}
	cmd.Stdout = os.Stdout
	cmd.Run()
}

func main() {
	rand.Seed(time.Now().UnixNano())
	size := 4
	if len(os.Args) > 1 {
		if v, err := strconv.Atoi(os.Args[1]); err == nil {
			size = v
			if size < 3 || size > 5 {
				fmt.Println("Размер должен быть от 3 до 5")
				os.Exit(1)
			}
		}
	}
	game := NewPuzzle(size, nil)
	game.shuffle(100)
	for !game.isSolved() {
		clearScreen()
		fmt.Println(colorize("ПЯТНАШКИ", bold))
		fmt.Println(colorize(fmt.Sprintf("Размер: %dx%d", size, size), blue))
		fmt.Println(game.render())
		elapsed := 0
		if !game.startTime.IsZero() {
			elapsed = int(time.Since(game.startTime).Seconds())
		}
		fmt.Println(colorize(fmt.Sprintf("Ходы: %d  Время: %d сек", game.moves, elapsed), yellow))
		fmt.Println(colorize("Управление: стрелки (WASD), R - перемешать, Q - выход", gray))
		reader := bufio.NewReader(os.Stdin)
		key, _ := reader.ReadString('\n')
		key = strings.TrimSpace(key)
		var moved bool
		switch key {
		case "q", "Q":
			fmt.Println("Выход.")
			return
		case "r", "R":
			game.shuffle(100)
			continue
		case "w", "W":
			moved = game.move(-1, 0)
		case "s", "S":
			moved = game.move(1, 0)
		case "a", "A":
			moved = game.move(0, -1)
		case "d", "D":
			moved = game.move(0, 1)
		default:
			fmt.Println("Неизвестная команда.")
			time.Sleep(500 * time.Millisecond)
			continue
		}
		if !moved && (key == "w" || key == "W" || key == "s" || key == "S" || key == "a" || key == "A" || key == "d" || key == "D") {
			fmt.Println("Невозможный ход.")
			time.Sleep(500 * time.Millisecond)
		}
	}
	clearScreen()
	fmt.Println(colorize("🎉 Поздравляем! Вы собрали головоломку!", green))
	elapsed := 0
	if !game.startTime.IsZero() {
		elapsed = int(time.Since(game.startTime).Seconds())
	}
	fmt.Println(colorize(fmt.Sprintf("Ходов: %d  Время: %d сек", game.moves, elapsed), yellow))
}

File: puzzle.py
# puzzle.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import sys
import os
import random
import time
import json
import argparse
from pathlib import Path

# ANSI colors
COLORS = {
    'reset': '\033[0m',
    'red': '\033[91m',
    'green': '\033[92m',
    'yellow': '\033[93m',
    'blue': '\033[94m',
    'magenta': '\033[95m',
    'cyan': '\033[96m',
    'white': '\033[97m',
    'gray': '\033[90m',
    'bold': '\033[1m'
}

def colorize(text, color):
    return f"{COLORS.get(color, '')}{text}{COLORS['reset']}"

# Конфигурация
RECORD_FILE = Path.home() / '.puzzle_records.json'

def load_records():
    if RECORD_FILE.exists():
        try:
            with open(RECORD_FILE, 'r') as f:
                return json.load(f)
        except:
            pass
    return {}

def save_records(records):
    with open(RECORD_FILE, 'w') as f:
        json.dump(records, f, indent=2)

class Puzzle:
    def __init__(self, size=4, state=None):
        self.size = size
        self.empty = (size-1, size-1)
        if state:
            self.board = [row[:] for row in state]
            # Найдём пустую
            for i in range(size):
                for j in range(size):
                    if self.board[i][j] == 0:
                        self.empty = (i, j)
        else:
            # Создаём решённое состояние
            self.board = [[i*size + j + 1 for j in range(size)] for i in range(size)]
            self.board[-1][-1] = 0
            self.empty = (size-1, size-1)
        self.moves = 0
        self.start_time = None
        self.size_str = f"{size}x{size}"

    def __str__(self):
        return self.render()

    def render(self):
        result = []
        result.append(colorize('┌' + '─' * (self.size * 4 - 1) + '┐', 'gray'))
        for i in range(self.size):
            row = []
            for j in range(self.size):
                val = self.board[i][j]
                if val == 0:
                    row.append('   ')
                else:
                    color = ['red', 'green', 'yellow', 'blue', 'magenta', 'cyan', 'white', 'gray'][val % 7]
                    row.append(colorize(f'{val:2} ', color))
            result.append(colorize('│', 'gray') + ''.join(row) + colorize('│', 'gray'))
        result.append(colorize('└' + '─' * (self.size * 4 - 1) + '┘', 'gray'))
        return '\n'.join(result)

    def is_solved(self):
        target = [[i*self.size + j + 1 for j in range(self.size)] for i in range(self.size)]
        target[-1][-1] = 0
        return self.board == target

    def is_solvable(self):
        # Преобразуем в список
        flat = [self.board[i][j] for i in range(self.size) for j in range(self.size) if self.board[i][j] != 0]
        inversions = 0
        for i in range(len(flat)):
            for j in range(i+1, len(flat)):
                if flat[i] > flat[j]:
                    inversions += 1
        if self.size % 2 == 1:
            return inversions % 2 == 0
        else:
            # Для чётного размера учитываем положение пустой
            empty_row = self.empty[0]
            if empty_row % 2 == 0:
                return inversions % 2 == 1
            else:
                return inversions % 2 == 0

    def shuffle(self, moves=100):
        # Делаем случайные ходы, чтобы перемешать
        dirs = [(0, 1), (0, -1), (1, 0), (-1, 0)]
        for _ in range(moves):
            r, c = self.empty
            random.shuffle(dirs)
            for dr, dc in dirs:
                nr, nc = r + dr, c + dc
                if 0 <= nr < self.size and 0 <= nc < self.size:
                    self.board[r][c], self.board[nr][nc] = self.board[nr][nc], self.board[r][c]
                    self.empty = (nr, nc)
                    break
        self.moves = 0
        self.start_time = None

    def move(self, dr, dc):
        r, c = self.empty
        nr, nc = r + dr, c + dc
        if 0 <= nr < self.size and 0 <= nc < self.size:
            self.board[r][c], self.board[nr][nc] = self.board[nr][nc], self.board[r][c]
            self.empty = (nr, nc)
            self.moves += 1
            if self.start_time is None:
                self.start_time = time.time()
            return True
        return False

def main():
    parser = argparse.ArgumentParser(description="Fifteen Puzzle – Пятнашки")
    parser.add_argument('size', nargs='?', type=int, default=4, help='Размер поля (3,4,5)')
    parser.add_argument('-s', '--seed', help='Строка с начальным состоянием через пробел')
    parser.add_argument('-r', '--record', action='store_true', help='Показать рекорд для данного размера')
    args = parser.parse_args()

    size = args.size
    if size < 3 or size > 5:
        print("Размер должен быть от 3 до 5")
        sys.exit(1)

    records = load_records()
    key = f"{size}x{size}"

    if args.record:
        if key in records:
            rec = records[key]
            print(colorize(f"Рекорд для {key}: {rec['moves']} ходов за {rec['time']} сек", 'green'))
        else:
            print(colorize(f"Нет рекорда для {key}", 'yellow'))
        sys.exit(0)

    # Создаём игру
    if args.seed:
        nums = list(map(int, args.seed.split()))
        if len(nums) != size*size:
            print("Неверное количество чисел")
            sys.exit(1)
        board = [nums[i*size:(i+1)*size] for i in range(size)]
        game = Puzzle(size, board)
    else:
        game = Puzzle(size)
        game.shuffle()

    # Проверка решаемости
    if not game.is_solvable():
        print("Эта комбинация нерешаема. Перемешиваем заново...")
        game = Puzzle(size)
        game.shuffle()

    # Цикл игры
    def clear_screen():
        os.system('clear' if os.name == 'posix' else 'cls')

    while not game.is_solved():
        clear_screen()
        print(colorize("ПЯТНАШКИ", 'bold'))
        print(colorize(f"Размер: {size}x{size}", 'blue'))
        print(game.render())
        elapsed = int(time.time() - game.start_time) if game.start_time else 0
        print(colorize(f"Ходы: {game.moves}  Время: {elapsed} сек", 'yellow'))
        print(colorize("Управление: стрелки (WASD), R - перемешать, Q - выход", 'gray'))
        try:
            key = input("Введите ход: ").strip().lower()
        except KeyboardInterrupt:
            print("\nВыход.")
            sys.exit(0)

        if key == 'q':
            print("Выход.")
            sys.exit(0)
        elif key == 'r':
            game.shuffle()
            continue
        elif key in ('up', 'w', '↑'):
            moved = game.move(-1, 0)
        elif key in ('down', 's', '↓'):
            moved = game.move(1, 0)
        elif key in ('left', 'a', '←'):
            moved = game.move(0, -1)
        elif key in ('right', 'd', '→'):
            moved = game.move(0, 1)
        else:
            print("Неизвестная команда.")
            time.sleep(0.5)
            continue
        if not moved:
            print("Невозможный ход.")
            time.sleep(0.5)

    # Победа
    clear_screen()
    print(colorize("🎉 Поздравляем! Вы собрали головоломку!", 'green'))
    elapsed = int(time.time() - game.start_time) if game.start_time else 0
    print(colorize(f"Ходов: {game.moves}  Время: {elapsed} сек", 'yellow'))
    # Обновление рекорда
    if key in records:
        if game.moves < records[key]['moves'] or (game.moves == records[key]['moves'] and elapsed < records[key]['time']):
            records[key] = {'moves': game.moves, 'time': elapsed}
            print(colorize("🏆 Новый рекорд!", 'green'))
        else:
            print(colorize(f"Текущий рекорд: {records[key]['moves']} ходов за {records[key]['time']} сек", 'blue'))
    else:
        records[key] = {'moves': game.moves, 'time': elapsed}
        print(colorize("🏆 Новый рекорд!", 'green'))
    save_records(records)

if __name__ == '__main__':
    try:
        main()
    except KeyboardInterrupt:
        print("\nВыход.")
        sys.exit(0)
